### DISCUSSION TITLE
Is it possible to get ctx object in withUrqlClient regardless of ssr setting?

### DISCUSSION THREAD

**Discussion by bcwatson22** [urql-graphql/urql#2004] [Q&A] [answered]
Hi, first of all cheers for making such a fantastic client!

I'm having an issue as I need to access `ctx` in the `withUrqlClient` hoc when using the setting `{ ssr: false }` (which I need to do as the project relies on the use of Next's SSG `getStaticProps` method in the page file). 

Essentially I need to know whether Next is currently in preview mode, as the URL will change based on `ctx.preview` which in turns grabs the unpublished data from the CMS. 

At the moment when I log out `ctx` I can only see `restoreData` and `extractData` funcs but not any of the other Next CTX props (`params`, `locale`, `preview` etc):

```javascript
const Home = () => (
  <Page query={query}
    slug="/" />
);

const getUrl = (ctx) => {

  console.log(ctx);

  return contentUrl;

};

export const getStaticProps = async ({ preview = false }) => {

  const ssrCache = ssrExchange({ isClient: false });
  const client = initUrqlClient({
    url: preview ? previewUrl : contentUrl,
    exchanges: [dedupExchange, cacheExchange, ssrCache, fetchExchange]
  }, false);

  await client?.query(query).toPromise();

  return {
    props: {
      urqlState: ssrCache.extractData()
    },
    revalidate: 600
  };

};

export default withUrqlClient(
  (ctx) => ({
    url: getUrl(ctx)
  }),
  { ssr: false }
)(Home);
```

Is this possible? Am I doing something wrong? 

Or is there an alternative way of structuring things in Next which will allow me to know whether it's currently in preview mode when wrapping my page component with `withUrqlClient`?

Thanks :)

**Comment by JoviDeCroock** [comment, marked as answer]
This is something that's lacking in Next 😅 when we use `ssr: true` we run through `getInitialProps` so we can easily inject `ctx` for you. The only other place where `ctx` is available is within the `getXProps` methods so you'd have to inject it into props from there. There's no way for us to provide you with `ctx` in `ssr: false` mode, we have no control over that if we aren't leveraging the `getInitialProps` method.

So what you could do is:

```js
  return {
    props: {
      urqlState: ssrCache.extractData(),
      preview
    },
    revalidate: 600
  };
```

And then you can use it in `props.preview` related discussion: https://github.com/FormidableLabs/urql/discussions/1973#discussioncomment-1374112

**Reply by bcwatson22** [reply]
Thanks @JoviDeCroock! How would I use that to dynamically switch out the `url` value used in `withUrqlClient`? The below is coming back undefined.

What's the `url` value in `withUrqlClient` actually used for? And does this always need to match what's used in `initUrqlClient`?

```javascript
const getUrl = (ctx, preview) => {

  console.log(preview);

  return contentUrl;

};

export const getStaticProps = async ({ preview = false }) => {

  ...

  return {
    props: {
      urqlState: ssrCache.extractData(),
      preview
    },
    revalidate: 600
  };

};

export default withUrqlClient(
  (ctx, preview) => ({
    url: getUrl(ctx, preview)
  }),
  { ssr: false }
)(Home);
```

**Reply by JoviDeCroock** [reply]
You can change the url both in `useQuery` and in `initUrqlClient` when you are using `getStaticProps` `initUrqlClient` will dictate the urql because that's the client you're using. For your render time the one in `withUrqlClient` will count, this is a confusion for me as well where these functions like `getStaticProps` are completely separated from the render-time.

**Reply by bcwatson22** [reply]
OK sound! So I can just use the non-preview URL in `withUrqlClient` which will be used at build time, and then the preview URL can be dynamic in `initUrqlClient`. 

Thanks so much for your help and rapid answers! :D

**Reply by JoviDeCroock** [reply]
That sounds right to me! No problem, glad I could help!
